plugins/modules/azure_rm_firewallpolicy.py

```diff
@@ -0,0 +1,437 @@
+        )
+
+        self.results = dict(
+            changed=False,
+            state=dict()
+        )
+
+        self.resource_group = None
+        self.name = None
+        self.state = None
+        self.location = None
+        self.base_policy = None
+        self.threat_intel_mode = None
+        self.threat_intel_whitelist = None
+        self.tags = None
+
+        super(AzureRMFirewallPolicy, self).__init__(self.module_arg_spec,
+                                                    supports_check_mode=True)
+
+    def exec_module(self, **kwargs):
+        for key in list(self.module_arg_spec.keys()) + ['tags']:
+            setattr(self, key, kwargs[key])
+
+        changed = False
+        results = dict()
+        firewall_policy_old = None
+        firewall_policy_new = None
+        update_ip_address = False
+        update_fqdns = False
+
+        # retrieve resource group to make sure it exists
+        resource_group = self.get_resource_group(self.resource_group)
+        if not self.location:
+            # Set default location
+            self.location = resource_group.location
+
+        self.location = normalize_location_name(self.location)
+
+        if self.base_policy:
+            base_policy = self.parse_resource_to_dict(self.base_policy)
+            self.base_policy = format_resource_id(val=base_policy['name'],
+                                                  subscription_id=base_policy['subscription_id'],
+                                                  namespace='Microsoft.Network',
+                                                  types='firewallPolicies',
+                                                  resource_group=base_policy['resource_group'])
+
+        try:
+            self.log('Fetching Firewall policy {0}'.format(self.name))
+            firewall_policy_old = self.network_client.firewall_policies.get(self.resource_group, self.name)
+            # serialize object into a dictionary
+            results = self.firewallpolicy_to_dict(firewall_policy_old)
+            if self.state == 'present':
+                changed = False
+                update_tags, results['tags'] = self.update_tags(results['tags'])
+                if update_tags:
+                    changed = True
+                self.tags = results['tags']
+                if self.base_policy is not None:
+                    if ('base_policy' not in results and self.base_policy != "") or \
+                            ('base_policy' in results and self.base_policy != results['base_policy']['id']):
+                        changed = True
+                        results['base_policy'] = self.base_policy
+                if self.threat_intel_mode is not None and \
+                        self.threat_intel_mode.lower() != results['threat_intel_mode'].lower():
+                    changed = True
+                    results['threat_intel_mode'] = self.threat_intel_mode
+                if self.threat_intel_whitelist:
```

Comment:
The processing logic here doesn't make sense. The recommendation is to compare the current parameter with the obtained result. If the resource does not exist, the parameter is the current configuration and is created directly. If the resource exists and the obtained configuration information contains the current parameters, then no update is made and idempotency holds. If the resource exists, the obtained configuration information does not contain the current configuration parameters. Will update and add new configurations. Thank you very much!

Reply:
You can also refer to the update logic of tags. Thank you very much!

Reply:
Sure @Fred-sun Thank you for suggestions! Let me work on it.

Reply:
Hi @Fred-sun, I have updated the logic for self.threat_intel_whitelist as suggested. Could you please check it once again? Thank you very much!